lib/elixir/lib/kernel.ex

```diff
@@ -795,11 +798,13 @@ defmodule Kernel do
 
       iex> trunc(5.4)
       5
+
       iex> trunc(5.99)
       5
 
   """
-  @spec trunc(number) :: integer
+  @spec trunc(arg) :: arg when arg: integer
```

Comment:
The rule I use is that when a type is used as a parameter to another type I treat it as a set of types, and when the variable is never a parameter (i.e. only an argument and/or returned value) I treat it is as the exact value. For me this makes types spec as explicit as possible.

I am unaware of a situation where OTP breaks this rule but tools can interpret it as they wish. The lists example does not break this rule as `A` and `B` are used as parameters.

```elixir
@spec call_then_double((() -> result)) :: result when result: integer | float
```
Breaks all the rules I think because the returned value is neither the exact value nor the same set of integers returned by the fun. For example the fun may return `4..6` but then function returns `8..12`.